vividus-plugin-email/src/main/java/org/vividus/bdd/email/service/ImapFetchService.java

```diff
@@ -0,0 +1,351 @@
+
+            if (filtered.length > 0)
+            {
+                fetchMessages(mailFolder, filtered, Set.of(Item.CONTENT_INFO));
+                return asMailMessages(List.of(filtered));
+            }
+            else
+            {
+                Waiter waiter = new Waiter(waitMode);
+                List<Message> output = interruptible(() -> waiter.wait(listener::getMessages, msgs -> !msgs.isEmpty()));
+                return asMailMessages(output);
+            }
+        }
+        catch (MessagingException | EmailMessageCreationException | InternalEmailFetchServiceException e)
+        {
+            throw new EmailFetchServiceException(e);
+        }
+    }
+
+    private List<EmailMessage> asMailMessages(List<Message> messages) throws EmailMessageCreationException
+    {
+        List<EmailMessage> emailMessages = new ArrayList<>(messages.size());
+        for (Message message : messages)
+        {
+            emailMessages.add(emailMessageFactory.create(message));
+        }
+        return emailMessages;
+    }
+
+    private Folder getFolder(Store store) throws MessagingException
+    {
+        store.connect();
+        return store.getFolder(this.folder);
+    }
+
+    private Message[] fetchMessages(Folder folder, Set<Item> fetchSettings) throws MessagingException
+    {
+        return fetchMessages(folder, folder.getMessages(), fetchSettings);
+    }
+
+    private Message[] fetchMessages(Folder folder, Message[] messages, Set<Item> fetchSettings)
+            throws MessagingException
+    {
+        FetchProfile profile = new FetchProfile();
+        fetchSettings.forEach(profile::add);
+        folder.fetch(messages, profile);
+        return messages;
+    }
+
+    private Map<String, String> asImapsProperties(Map<String, String> properties)
+    {
+        return properties.entrySet().stream()
+                .collect(Collectors.toMap(e -> "mail." + PROTOCOL + "." + e.getKey(), Map.Entry::getValue));
+    }
+
+    private <T> T interruptible(CheckedSupplier<T, InterruptedException> supplier)
+    {
+        try
+        {
+            return supplier.get();
+        }
+        catch (InterruptedException e)
+        {
+            Thread.currentThread().interrupt();
+            throw new IllegalStateException(e);
+        }
+    }
+
+    private static final class PredicateSearchTerm extends SearchTerm
+    {
+        private static final long serialVersionUID = 1163386376061414046L;
+
+        @SuppressFBWarnings("SE_TRANSIENT_FIELD_NOT_RESTORED")
+        private final transient List<CheckedPredicate<Message, MessagingException>> messageFilters;
+
+        private PredicateSearchTerm(List<CheckedPredicate<Message, MessagingException>> messageFilters)
+        {
+            this.messageFilters = messageFilters;
+        }
+
+        @Override
+        public boolean match(Message msg)
+        {
+            for (CheckedPredicate<Message, MessagingException> filter : messageFilters)
```

Comment:
Here you are testing the message, and what if the message is not the one you're waiting but one of the search criteria leads to an exception.
 Does it mean inability to interact with the server and process other messages?

Reply:
when looking at MessagingException hierarchy I can assume that it can be either MailConnectException or MessageRemovedException or MethodNotSupportedException. I will add handling of MessageRemovedException